src/main/scala/firrtl/graph/DiGraph.scala

```diff
@@ -1,102 +1,59 @@
 package firrtl.graph
 
-import scala.collection.immutable.{Set, Map, HashSet, HashMap}
-import scala.collection.mutable
-import scala.collection.mutable.MultiMap
-
-/** Represents common behavior of all directed graphs */
-trait DiGraphLike[T] {
-  /** Check whether the graph contains vertex v */
-  def contains(v: T): Boolean
-
-  /** Get all vertices in the graph
-    * @return a Set[T] of all vertices in the graph
-    */
-  def getVertices: collection.Set[T]
-
-  /** Get all edges of a node
-    * @param v the specified node
-    * @return a Set[T] of all vertices that v has edges to
-    */
-  def getEdges(v: T): collection.Set[T]
-}
-
-/** A class to represent a mutable directed graph with nodes of type T
-  * 
-  * @constructor Create a new graph with the provided edge data
-  * @param edges a mutable.MultiMap[T,T] of edge data
-  * 
-  * For the edge data MultiMap, the values associated with each vertex
-  * u in the graph are the vertices with inedges from u
-  */
-class MutableDiGraph[T](
-  private[graph] val edgeData: MultiMap[T,T] =
-    new mutable.HashMap[T, mutable.Set[T]] with MultiMap[T, T]) extends DiGraphLike[T] {
-
-  // Inherited methods from DiGraphLike
-  def contains(v: T) = edgeData.contains(v)
-  def getVertices = edgeData.keySet
-  def getEdges(v: T) = edgeData(v)
-
-  /** Add vertex v to the graph
-    * @return v, the added vertex
-    */
-  def addVertex(v: T): T = {
-    edgeData.getOrElseUpdate(v,new mutable.HashSet[T])
-    v
-  }
-
-  /** Add edge (u,v) to the graph */
-  def addEdge(u: T, v: T) = {
-    // Add v to keys to maintain invariant that all vertices are keys
-    // of edge data
-    edgeData.getOrElseUpdate(v, new mutable.HashSet[T])
-    edgeData.addBinding(u,v)
-  }
-}
+import scala.collection._
+import scala.collection.mutable.{LinkedHashSet, LinkedHashMap}
 
 /** A companion to create immutable DiGraphs from mutable data */
 object DiGraph {
   /** Create a DiGraph from a MutableDigraph, representing the same graph */
-  def apply[T](mdg: MutableDiGraph[T]): DiGraph[T] =
-    new DiGraph((mdg.edgeData mapValues { _.toSet }).toMap[T, Set[T]])
-
-  /** Create a DiGraph from a MultiMap[T] of edge data */
-  def apply[T](edgeData: MultiMap[T,T]): DiGraph[T] =
-    new DiGraph((edgeData mapValues { _.toSet }).toMap[T, Set[T]])
+  def apply[T](mdg: MutableDiGraph[T]): DiGraph[T] = mdg
 
   /** Create a DiGraph from a Map[T,Set[T]] of edge data */
-  def apply[T](edgeData: Map[T,Set[T]]) = new DiGraph(edgeData)
+  def apply[T](edgeData: Map[T,Set[T]]) = {
```

Comment:
If the edgeData already is a LinkedHashMap, would it be reasonable to shortcut to just creating the DiGraph? Could be another apply method.

I'm writing code to use this I imagine for anyone who cares about determinism, this will be a common use case?

Reply:
The factory method you're proposing is dangerous because it creates an 'immutable' digraph that is backed by a mutable collection that the user can modify.

Personally, I would use it if it existed, but I don't know the convention around that kind of mutation hole.

Reply:
Ah good point, so I guess the right pattern here is to create a Mutable DiGraph and mutate that rather than creating the original LinkedHashMap?

Reply:
Then create an immutable DiGraph from the mutable one?

Reply:
That was the original intent